azkaban-web-server/src/main/java/cloudflow/services/ProjectServiceImpl.java

```diff
@@ -0,0 +1,139 @@
+    int spaceIdInt = Integer.parseInt(spaceId);
+    Optional<Space> space = spaceDao.get(spaceIdInt);
+    if (!space.isPresent()) {
+      String errorMsg = String.format("Space id: %s passed by user while creating project: %s "
+          + "does not exist", spaceId, project.getName());
+      throw new CloudFlowValidationException(errorMsg);
+    }
+  }
+
+
+  public void validateInput(Project project) {
+    /* Check required fields are present */
+    requireNonNull(project.getSpaceId(), "Space id is null");
+    requireNonNull(project.getName(), "Name is null");
+
+    validateSpaceId(project);
+
+    /* Check all the fields that Azkaban will set are empty */
+    if (project.getId() != null || project.getLatestVersion() != null ||
+        project.getCreatedByUser() != null || project.getLastModifiedByUser() != null ||
+        project.getCreatedOn() != 0 || project.getLastModifiedOn() != 0) {
+      String errorMsg = String.format("A forbidden field was specified in request body for "
+          + "project: %s", project.getName());
+      log.error(errorMsg);
+      throw new CloudFlowValidationException(errorMsg);
+    }
+
+    /* Check if admins list has any duplicates */
+    Set<String> adminNames = new HashSet<String>();
+    for (String name : project.getAdmins()) {
+      if (adminNames.contains(name)) {
+        String errorMsg = String.format("Duplicate userid: %s in admins list for project: %s",
+            name, project.getName());
+        log.error(errorMsg);
+        throw new CloudFlowValidationException(errorMsg);
+      } else {
+        adminNames.add(name);
+      }
+    }
+  }
+
+  @Override
+  public String createProject(Project project, User user) {
+    validateInput(project);
+    return this.projectDao.create(project, user);
+  }
+
+  @Override
+  public Project getProject(final String projectId) {
+
+    Optional<Project> project = projectDao.get(projectId);
+    /* Not the best exception to throw but serves the use-case here.
+     */
+    if (!project.isPresent()) {
+      String errorMsg = String.format("Project record doesn't exist for id: %s", projectId);
+      log.error(errorMsg);
+      throw new CloudFlowNotFoundException(errorMsg);
+    }
+    return project.get();
+  }
+
+  @Override
+  public List<Project> getAllProjects(User user) {
+    List<Project> projects = new ArrayList<Project>();
+
+    /* Get all projects created or modified by user*/
+    projects.addAll(projectDao.getAll(user));
+
+    /* Get all projects for which user is an admin */
+    List<String> projectIds = projectAdminDao.findProjectIdsByAdmin(user.getUserId());
+    for (String projectId : projectIds) {
+      try {
+        Project project = getProject(projectId);
+        /* Add project only if user did not create/modify the project, otherwise the project will
+         already be added when we get all projects for a user above. */
+        if (!projects.contains(project)) {
+          projects.add(project);
+        }
```

Comment:
No need for the question if we use a Set instead. It's a noop trying to add an element to a set that's already there.

Reply:
It would be worthwhile to look into changing the return type of getAll() methods in all services to set instead of a list.